install_files/ansible-base/roles/grsecurity/tasks/paxctl.yml

```diff
@@ -0,0 +1,36 @@
+---
+- name: Install paxctl.
+  apt:
+    pkg: paxctl
+    state: present
+  tags:
+    - apt
+
+- name: Check paxctl headers on grub binaries.
+  command: paxctl -vQ {{ item }}
+  # Read-only task, so don't report changed
+  changed_when: false
+  # The first time this task is run, the grub binaries won't have
+  # their headers converted yet, which will cause paxctl to exit 1.
+  # We'll catch that error and respond accordingly in the next task.
+  failed_when: false
+  register: paxctl_grub_header_check
+  with_items:
+    - /usr/sbin/grub-probe
+    - /usr/sbin/grub-mkdevicemap
+
+- name: Adjust paxctl headers on grub binaries.
+  command: paxctl -Cpm {{ item.item }}
+  with_items: paxctl_grub_header_check.results
+  # The desired flags should include:
+  #   - p NOPAGEEXEC
+  #   - m NOMPROTECT
+  #   - x NORANDEXEC
+  #   - e NOEMUTRAMP
+  # After running `paxctl -Cpm` in VirtualBox, all four of those flags are set.
+  # However, the current paxctl command only enforces p and m.
+  # The x and e flags are present by default, but can be removed via
+  # `paxctl -z <binary>`. TODO: We may want to update the command
+  # to set -Cpmxe explicitly. Warrants further testing on hardware.
```

Comment:
We should probably revisit these Pax flags generally. My usual stance is to try to avoid removing any of grsecurity's protections unless it is absolutely necessary. Sometimes it is necessary, for example if the program needs to be rewritten to understand the error code grsec returns from an mmap for WX, or if it uses trampolined functions, etc.

The Pax rules for GRUB were cargo-culted from an old blog post about getting grsec working on Ubuntu, IIRC. It is possible that grub has since been updated such that they are no longer necessary, in which case we should allow it to take advantage of the full set of protections afforded by grsecurity.